packages/channel/src/buffers.ts

```diff
@@ -1,95 +1,102 @@
-export interface ChannelBuffer<T> {
-  full: boolean;
-  empty: boolean;
-  add(value: T): void;
-  remove(): T;
+import {
+  CannotReadFromEmptyBufferError,
+  CannotWriteToFullBufferError,
+  InvalidBufferCapacityError,
+} from "./errors";
+
+export abstract class ChannelBuffer<T> {
+  public abstract get full(): boolean;
+  public abstract get empty(): boolean;
+
+  public abstract add(value: T): void;
+  public abstract remove(): T;
+
+  constructor(protected capacity: number) {
+    if (capacity < 0) {
+      throw new InvalidBufferCapacityError(capacity);
```

Comment:
I don’t see the benefit of having a custom error class here over a `RangeError`. No one should be catching `InvalidBufferCapacityError` insofar as it is strictly a programmer error which is thrown on initialization.